src/services/codefixes/inferFromUsage.ts

```diff
@@ -73,7 +73,11 @@ namespace ts.codefix {
                     const type = inferTypeForVariableFromUsage(parent.name, program, cancellationToken);
                     const typeNode = type && getTypeNodeIfAccessible(type, parent, program, host);
                     if (typeNode) {
-                        changes.tryInsertJSDocType(sourceFile, parent, typeNode);
+                        // Note that the codefix will never fire with an existing `@type` tag, so there is no need to merge tags
+                        const typeTags = createMap<JSDocTypeTag>();
+                        const typeTag = createJSDocTypeTag(typeNode, /*comment*/ "");
+                        typeTags.set("UNUSED", typeTag);

```

Comment:
Well, there's a lot of shared behaviour here (see the analysis below), so I think the most reasonable change is to make `insertMergedJSDoc` take a `T | Map<T>` instead of just `Map<T>`. I tried that and give it at the bottom of this comment.

There are three behaviours here:

1. Always add a `@type` tag, which you see here. That's because we will only ever run the codefix if there is no type tag, and the only thing we'll add is a type tag. However, we still want to merge other tags as well as top-level comments.
2. Add or merge a single `@type` or `@return` tag, where there might be an existing tag with no type -- this applies to `@return` only, since a typeless `@type` is an error. We want merging here too.
3. Add or merge a number of parameter tags, distinguished by name. We want merging here too.

So the top-level merging code needs to be shared across all three scenarios. (2) and (3) need per-tag merging as well, and when a tag has been merged, it needs to not appear as a new tag, but replace the old tag.

Here's my attempt at getting rid of the map. Note that tracking already-merged is duplicated between the single-tag (2) and multi-tag (3) cases.

```suggestion
    function insertMergedJSDoc<T extends JSDocTag>(
        syntheticTags: T | Map<T>,
        changes: textChanges.ChangeTracker,
        sourceFile: SourceFile,
        parent: HasJSDoc,
        getKey: (tag: JSDocTag) => string | false,
        merge: (synthetic: T, existing: T) => void) {
        const comments = [];
        const tags: JSDocTag[] = [];
        let used = false;
        if (hasJSDocNodes(parent)) {
            for (const jsdoc of parent.jsDoc!) {
                if (jsdoc.comment) {
                    comments.push(jsdoc.comment);
                }
                if (jsdoc.tags) {
                    for (const existing of jsdoc.tags) {
                        const key = getKey(existing);
                        const synthetic = 'atToken' in syntheticTags ?
                            syntheticTags.kind === existing.kind && !used && syntheticTags :
                            key && syntheticTags.get(key);
                        if (synthetic) {
                            merge(synthetic, existing as T);
                            tags.push(synthetic);
                            if ('atToken' in syntheticTags) {
                                used = true;
                            }
                            else {
                                syntheticTags.delete(key as string);
                            }
                        }
                        else {
                            tags.push(existing);
                        }
                    }
                }
            }
        }
        if ('atToken' in syntheticTags) {
            if (!used) {
                tags.push(syntheticTags);
            }
        }
        else {
            syntheticTags.forEach(v => tags.push(v));
        }
        const tag = createJSDocComment(comments.join("\n"), createNodeArray(tags));
        changes.insertJsdocCommentBefore(sourceFile, parent, tag);
    }
```

Reply:
Also note that the keys of `syntheticTags` are only ever accessed via the result of `getKey`, so when `getKey` returns "TAG" and only "TAG", it is guaranteed never to clash with any other names.